scylla-cdc-printer/src/printer.rs

```diff
@@ -0,0 +1,246 @@
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use anyhow;
+use chrono::Duration;
+use core::time;
+use scylla::Session;
+use scylla::frame::response::result::Row;
+use tokio::time::sleep;
+
+use scylla_cdc::cdc_types::GenerationTimestamp;
+use scylla_cdc::reader::StreamReader;
+use scylla_cdc::stream_generations::GenerationFetcher;
+
+struct CDCLogPrinter {
+    session: Arc<Session>,
+    start_timestamp: Duration,
+    end_timestamp: Duration,
+    sleep_interval: time::Duration,
+    window_size: Duration,
+    safety_interval: Duration,
+    readers: Vec<Arc<StreamReader>>,
+}
+
+impl CDCLogPrinter {
+    fn new(
+        session: &Arc<Session>,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+    ) -> CDCLogPrinter {
+        CDCLogPrinter {
+            session: Arc::clone(session),
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            readers: vec![],
+        }
+    }
+
+    pub async fn run(
+        &mut self,
+        keyspace: String,
+        table_name: String,
+        mut after_fetch_callback: impl FnMut(Row),
+    ) -> anyhow::Result<()> {
+        let fetcher = &GenerationFetcher::new(&self.session);
+        let mut generation = self.get_first_generation(fetcher).await?;
+
+        loop {
+            self.readers = vec![];
+            for stream_id in fetcher
+                .fetch_stream_ids(&generation)
+                .await?
+                .iter()
+                .flatten()
+            {
+                let reader = StreamReader::new(
+                    &self.session,
+                    vec![stream_id.clone()],
+                    self.start_timestamp,
+                    self.window_size,
+                    self.safety_interval,
+                    self.sleep_interval,
+                );
+                self.readers.push(Arc::new(reader));
+            }
+
+            let handles = self.readers.iter().map(|reader| {
+                tokio::spawn(async move {
```

Comment:
You are using an `async move` block here - because it is marked as `move`, it moves inside itself _all_ variables from the outer scope which are used in the block. Currently, there is going to be a problem with every one of them:

- `reader` - it has type `&Arc<StreamReader>` and is a reference to one of the readers in the `CDCLogPrinter`. Because of it, the async block needs to be destroyed before the Arc pointer to by the reference, otherwise there will be a dangling reference. The futures running in tokio tasks are required to have static lifetimes and rustc will complain because the captured reference will make the future (async block) non-static.
- `keyspace`. `table_name`, `after_fetch_callback` - the closure at line 73 captures those variables from outer scope by reference, but the async block tries to move them in. It won't work.

The solution is to clone those variables _before_ moving them to the async block:

```rust
let handles = self.readers.iter().map(|reader| {
    let reader = Arc::clone(reader);
    // The next line works provided that the fetch callback is cloneable
    // This might or might not be something that we want
    let cb = after_fetch_callback.clone();
    let keyspace = keyspace.clone();
    let table_name = table_name.clone();
    tokio::spawn(async move {
        reader
            .fetch_cdc(keyspace.clone(), table_name.clone(), cb)
            .await
    })
});
```